### PR TITLE
[PHP-Client] Allow Content-Type merge-match+json for encoding

### PR DESCRIPTION
**Pull request by frickelbruder**
Some APIs offer a PATCH-Request method for Entities and use "application/merge-patch+json" as Content-Type. This was not detected as json_encode and ObjectSerializer target and thus resultet in an Exception,

```
Nyholm\Psr7\Request::withBody(): Argument #1 ($body) must be of type Psr\Http\Message\StreamInterface, _XXX__ given, called in XXX
```

This PR fixes this.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.6.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@jebentier , @dkarlovi , @mandrean , @jfastnacht , @ybelenko , @renepardon

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/php/libraries/psr-18/api.mustache [hunk 1]
```diff
@@ -93,6 +93,14 @@ use function sprintf;
...
      */
     protected $streamFactory;
+    
+    /**
+     * @var string[]
+     */
+    public const JSON_FORMATS = [
+        'application/merge-patch+json', 
+        'application/json',
+    ];
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/php/libraries/psr-18/api.mustache:103]: what about using a regular expression to detect json related MIME type?

https://github.com/OpenAPITools/openapi-generator/pull/19317/files#diff-c1d7ea3a1de60effaa5908cee1be9a7e2fd420bdb75a2e2a28a6c3290380192dR415

Comment (frickelbruder) [reply modules/openapi-generator/src/main/resources/php/libraries/psr-18/api.mustache:103]: Sounds good, I updated the code.

File: modules/openapi-generator/src/main/resources/php/libraries/psr-18/api.mustache [hunk 2]
```diff
@@ -93,7 +93,7 @@ use function sprintf;
      * @var StreamFactoryInterface
      */
     protected $streamFactory;
-
+
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/php/libraries/psr-18/api.mustache:96]: can these spaces in the blank line be removed?

File: samples/client/petstore/php/psr-18/lib/Api/DefaultApi.php [hunk 3]
```diff
@@ -464,4 +464,9 @@ private function createUri(
 
         return $uri;
     }
+
+    private function bodyShouldBeEncoded(string $contentType): bool 
+    {        
+        return preg_match('~^application/(json|[\w!#$&.+-^_]+\+json)\s*(;|$)~', $contentType) === 1;
+    }
```
Comment (wing328) [inline samples/client/petstore/php/psr-18/lib/Api/DefaultApi.php:471]: what about adding this function in the api client class instead?

can we name this method as isJsonMime?

minor suggestion: add a docstring explaining what this function does

Comment (frickelbruder) [reply samples/client/petstore/php/psr-18/lib/Api/DefaultApi.php:471]: I am not sure, what you mean by "what about adding this function in the api client class instead?"
As this is a method used throughout the code and if not working properly results in Exceptions being thrown, I think it is quite important.
What I could think of, is implementing a separate Class "JsonDetector" or anything with a single method, which could be injected via consturctor DI. Would this be a good solution or are you thinking of anything else?

Comment (wing328) [reply samples/client/petstore/php/psr-18/lib/Api/DefaultApi.php:471]: sorry for the confusion, api client class (ApiClient.php) is in client generated by php-dt generator but not php client generator.

shall we put the function in HeaderSelector class instead which has something similar already?

https://github.com/OpenAPITools/openapi-generator/blob/11a483ef92eff347b826d18398191d97406a2d68/samples/client/petstore/php/psr-18/lib/HeaderSelector.php#L88

Comment (frickelbruder) [reply samples/client/petstore/php/psr-18/lib/Api/DefaultApi.php]: Did you have something like this in mind?

**Comment by wing328** [comment]
all tests passed locally:
```
[INFO]
[INFO] --- exec:1.2.1:exec (bundle-test) @ PhpPetstoreOAS3Tests ---
PHPUnit 9.6.19 by Sebastian Bergmann and contributors.

...............................................................  63 / 222 ( 28%)
............................................................... 126 / 222 ( 56%)
............................................................... 189 / 222 ( 85%)
.................................                               222 / 222 (100%)

Time: 00:01.709, Memory: 12.00 MB

OK (222 tests, 384 assertions)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  10.721 s
[INFO] Finished at: 2024-10-03T20:52:32+08:00
[INFO] ------------------------------------------------------------------------
[INFO] 3 goals, 3 executed
```

thanks again for the PR

**Pull request merged**